### PR TITLE
CMW-2200: Response Package

### PR DESCRIPTION
**Pull request by zlarsen**
## Description

image

## Steps to Test
1. Run `mix test`, and ensure tests pass
2. Review code and see if any functions that you would like in this version are missing

## Screenshots
image

### CODE REVIEW

File: lib/changeset.ex [hunk 1]
```diff
@@ -0,0 +1,162 @@
+defmodule Solicit.Changeset do
+  @moduledoc """
+  Generic functions for working with changesets that are not packaged with Ecto.
+  """
+
+  @doc """
+  A module encapsulating the translation layer between a changeset error and the JSON error shape returned in an API response.
+  """
+  @spec code_and_description(map()) :: map()
+  def code_and_description({"can't be blank", _}),
```
Comment (bjyoungblood) [inline lib/changeset.ex:10]: Pattern matching on the description text seems kind of brittle to me. Is there a reason not to match on the `validation` key from the tuple's second member?

**Review by Scott-Christianson — APPROVED** [review]
Good stuff, dude! Can't wait for this!

File: lib/changeset.ex [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+      do: {:invalid_value, "This property should have at most #{count} characters."}
+
+  # V1 of mobile apps respond to this error
+  def code_and_description({"duplicate code", _}),
+    do: {:duplicate_code, "Code already in use."}
+
+  # If a custom validation error was created with an error_code attached,
+  # use that error_code (and the error message as the description)
+  def code_and_description({description, [error_code: error_code]}),
+    do: {error_code, description}
```
Comment (bjyoungblood) [inline lib/changeset.ex:159]: Where do we draw the line between what goes in this package and what doesn't? Do we have to add something here every time we add custom validation in any project? What happens if we forget?

File: md/response.md [hunk 3]
```diff
@@ -0,0 +1,194 @@
+# Response
```
Comment (bjyoungblood) [inline md/response.md:1]: Docs are ❤️ , but `@doc` attributes >>>> a random markdown file. 😉

File: lib/response.ex [hunk 4]
```diff
@@ -0,0 +1,301 @@
+defmodule Solicit.Response do
```
Comment (bjyoungblood) [inline lib/response.ex:1]: Thoughts on consolidating the package's public API (this looks like the majority of what consumers will be using) to the top-level `Solicit` module?

@mattludwigs posted this (warning: long) article on Elixir API design on Slack a couple months ago https://aaronrenner.io/2019/09/18/application-layering-a-pattern-for-extensible-elixir-application-design.html

File: mix.exs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+  # Run "mix help deps" to learn about dependencies.
+  defp deps do
+    [
+      {:ecto, "~> 3.4"},
+      {:ex_doc, "~> 0.16", only: :dev, runtime: false},
+      {:jason, "~> 1.0"},
+      {:phoenix_ecto, "~> 4.1.0"},
+      {:phoenix, "~> 1.4"},
+      {:plug_cowboy, "~> 1.0"},
```
Comment (bjyoungblood) [inline mix.exs:27]:
```suggestion
      {:plug_cowboy, ">= 1.0.0"},
```

We have projects that use 2.x

File: lib/changeset.ex (hunk 1 above)
Comment (zlarsen) [reply lib/changeset.ex]: That's smart. I updated this one to match on the `validation` key. I couldn't find the `validation` key for a few others so they still match to `{"blah", _}`

File: lib/changeset.ex (hunk 2 above)
Comment (zlarsen) [reply lib/changeset.ex:159]: That's a really good point. I think we should go over this in the API Squad meeting on Monday.

File: md/response.md (hunk 3 above)
Comment (zlarsen) [reply md/response.md:1]: I agree! Added base `@doc` attributes as well.

Thank you!

File: lib/response.ex (hunk 4 above)
Comment (zlarsen) [reply lib/response.ex:1]: I think that could be a great API design. I think it may be worthwhile to get opinions on Monday as well.

**Pull request merged**